enhancements/olm/olm-managed-operator-metrics.md

```diff
@@ -0,0 +1,145 @@
+
+#### Telemeter Requirements
+
+Once OLM is capable of managing `ServiceMonitor` and `PrometheusRule` objects along with the lifecycle of an operator, OLM must allow Red Hat owned operators to take advantage of Telemeter. Before moving forward, let's review the requirements that must be met for Telemeter to scrape an application's metrics:
+
+1. The application must be deployed in a namespace that is both prefixed with `openshift-` and labeled with `openshift.io/cluster-monitoring=true`
+2. A [ServiceMonitor object](https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/getting-started.md#related-resources) must be created in the namespace mentioned above and point to the application's metrics endpoint
+3. A role/rolebinding must be created that provides the Prometheus Operator ServiceAccount with the appropriate RBAC priviledges to discover the `ServiceMonitors` and `PrometheusRules` objects in the newly created namespace.
+
+Both requirements 1 and 2 are fully met by the changes proposed above leaving our team with the duty to fulfill requirement 3.
+
+#### Fulfilling Requirement 3: Updating the Prometheus Operator ServiceAccount
+
+In order for the Prometheus Operator to recognize take advantage of this new feature, its ServiceAccount must have the proper RBAC permissions to view events on the `ServiceMonitors` and `PrometheusRules` objects in the operator's namespace.
+
+On upstream clusters, there is no way for OLM to identify which Prometheus Operator ServiceAccount should be granted the RBAC priviledges. As such, the cluster admin will need to update the correct Prometheus Operator's OperatorGroup to include the correct namespace.
+
+On OpenShift cluster, OLM can simply update the default Prometheus instance with the correct RBAC priviledges when the `openshift.io/cluster-monitoring=true` label is present on the namespace.
+
+#### Risks and Mitigations
+
+It is possible that the Prometheus team will not allow OLM managed operators to use the default prometheus operator to record metrics. If they are concerned about OLM managed operators overloading the prometheus instance, one solution that addresses this concern includes creating a new operator that is responsible for adding the `openshift.io/cluster-monitoring=true` label to namespaces and for generating the Telemter RBAC. This operator would:
+
+- Be deployed in the `openshift-monitoring` namespace
+- Have a list of approved OLM managed operators that may report metrics to Telemeter
+- Watch for events on namespaces
+- If an event occurs on a namespace that starts with `openshift-` and the namespace only includes valid operators, add the `openshift.io/cluster-monitoring=true` annotation and create RBAC that allows the Telemeter service account to watch for events in the namespace.
+
+Alternatively, we could only add the `openshift.io/cluster-monitoring=true` on operators that originated from the `redhat-operators` and `certified-operators` CatalogSources.
```

Comment:
Yes this was the plan we agreed on with @shawn-hurley. Only redhat operators, not the certified ones, are allowed to be deployed in openshift ns that have `openshift.io/cluster-monitoring=true`, because we consider that to be openshift workload.

Reply:
I don't think that we can limit it to just RedHat operators, and hence this extra bit in the design proposal about another option that the Monitoring team could use. 

We have tech-preview/developer-preview operators in the `Community Catalog` that I would think should be able to register for the monitoring. This alone means that we need something more robust.

Reply:
> We have tech-preview/developer-preview operators in the Community Catalog that I would think should be able to register for the monitoring. This alone means that we need something more robust.

But those then fall under user monitoring, no? This can be already done from 4.3 onwards via tech preview and will be GA in a couple of releases.

Reply:
That's definitely the primary target, but it would be nice if we had an option for the time when "hey this is a certified operator but we really need to see telemetry data for it" happens.

Reply:
@LiliC Does telemeter ever scrape "user monitoring" instances of prometheus?

Reply:
Any metrics are possible to be sent via telemetry they can be user monitoring or openshift workload metrics. But I am not sure we should be allowing non openshift operators to send metrics via telemetry. This is something we need to discuss further, outside of this PR.

Reply:
I don't think that we are advocating one way or the other here for either option. What we are advocating for is a robust enough solution to solve the use cases that people have today. This includes folks in community operators who should be able to use this. I think that the right thing is adding some guardrails around the label as we did discuss, and it seems that this net-new control loop should be able to handle it and that the monitoring team should be responsible for the operators that are able to be added to the workload monitoring as the OLM team does not have the experience/knowledge to make that decision.

Reply:
the primary concern/risk around a component that can contribute metrics is that any component that can contribute a metric (regardless of whether it goes to telemeter) can stomp on metrics contributed by other components.  Depending what those metrics are used for, that can impact billing/chargeback, alerting, etc.  (if it's a metric that goes back to telemeter, it's even worse).

So the decision point is basically, "do we trust this component not to do bad things".  Either we enforce that (by saying only openshift-* things and only our stuff goes in openshift-*), or we make the cluster-admin make that decision (by saying any operator can ask to be scraped and the cluster-admin can decide whether or not to install it(or whether or not to allow it to be scraped).

I would expect us to be able to trust certified operators, i would not expect us to be able to trust community operators, and i'd hate to put this decision in the hands of a cluster-admin, especially since it can affect our telemeter data.

The secondary concern would be a component that floods prometheus with metric data.  Again, we can probably trust certified operators, we can't trust community operators.

Reply:
I don't think we can trust certified operators.

I think for this proposal to work, we have to have a fairness algorithm.  I don't think any other solution is workable or worth the effort, because all these workarounds are just hacks on trust, but if we can reliably balance available capacity fairly among all operators and throttle misbehavers, we have no issue.  That also improves product reliability in general and is a straight win.

Reply:
Discussed this with @shawn-hurley and since we need a path forward for tech preview operators that live in the community catalog, and we need it before user workload monitoring goes GA, the plan is as follows:

1) allow community operators to request to be in the openshift-* namespace
2) we will enforce on the community operator repo, that only operators we trust get to request that (we review all the PRs but we'll need some automation/validation there)
3) the UI bits will need to warn the admin of any operator that is requesting to have metrics scraped out of the openshift-* namespace, except for operators that come out of the redhat +certified catalogs.

Reply:
Do we have a design for scrape fairness?

Reply:
i'm less worried about the scraper being overloaded and more worried about a bad operator creating a metric w/ the same name as our core metrics.

Reply:
@s-urbaniak @LiliC we should discuss the need/plans for a scraping fairness/abuse-protection system in prometheus, maybe in the next monitoring arch call.

Reply:
cc @brancz on the fairness algorithm initiative.

Reply:
I'm also not worried about scraping being overloaded, I am worried about overloading the time-series database and security/reliability/durability of the data.

There are some amount of things we can do about limiting the load on the time-series database, for example it's perfectly possible to configure a limit of time-series per target in a ServiceMonitor, and we could ensure there is an upper limit to this.

For technical reasons we currently have to allow honoring labels in the cluster monitoring Prometheus setup (this is disabled in user workload monitoring). This is what makes ingesting untrusted data dangerous, as this gives targets the power to decide which time-series to append data to, as opposed to the decision being 100% up to the Prometheus server. We have been working on getting rid of this eventually but it's a long story largely because of the shape of the data we ingest into cluster monitoring.

I think what Clayton captured [here](https://github.com/openshift/enhancements/pull/152#discussion_r358331578) is the crooks of it: how do we get very high quality, and high impact metrics, and I also feel OLM is perfectly capable of exposing these without Prometheus having to scrape the operator to begin with. With these types of metrics (essentially 3 metrics per operator) the cardinality is low enough that we could always collect them from OLM and we send the topk(30) via telemetry.

Reply:
We discussed this within the team with @s-urbaniak and @brancz and this is our monitoring perspective on the topic: 
1. OLM should expose the basic USE(D) metrics for every operator installed via OLM. **U**tilisation, **S**aturation, **E**rrors and **D**uration should be exposed by OLM for any operator, this could even include community ones, as the metric source at least is trusted. For example: install errors, number of custom resources created,...
2. Only Red Hat OLM installed operators can expose metrics in openshift-* namespace, so they can fall under the openshift cluster monitoring. This is to prevent any security and multi-tenancy problems.
3. Red Hat operators must be enabled in the e2e origin, even optional ones, to prevent high cardinality and reliability guarantees and must follow the regular cadence of OpenShift release cycle. This ensures that any changes impacting in-cluster-monitoring are encapsulated within official OpenShift releases as there are tests (some existing, some in progress) to verify that there is no cardinality explosion and overall cardinality is within reason.
4. This has been considered a low priority by our PM (even this consulting work), so if OLM wants to prioritise this, they need to speak with Christian about product decisions, as we do not have time to sheppard this for the next few releases.

Reply:
I don't think that this is possible. We have partners that need to expose info, and this also needs to work in a disconnected cluster (which means we can not lock down to a catalog source). 

We need to come up with a middle ground solution until user-workload monitoring is finished. 

OLM exposing metrics for every operator is being tracked in a different proposal enhancement for now.

Reply:
> We need to come up with a middle ground solution until user-workload monitoring is GA.

Until then the OLM installed prometheus-operator can be used.

Reply:
> We have partners that need to expose info

Not sure what this means, can you provide more details on this?

Reply:
Partners need the ability to go through the process to expose their metrics via telemeter. IIUC the prometheus-operator does not grant Users the ability to do this correct?

Reply:
> Partners need the ability to go through the process to expose their metrics via telemeter.

Why would partners want to send metrics via telemetry? To quote @smarterclayton "I don't think we can trust certified operators."

Reply:
Also the metrics sending via telemetry should be discussed outside of this PR as we agreed before.